certcache.go

```diff
@@ -56,3 +108,245 @@ func (this *CertCache) ServeHTTP(resp http.ResponseWriter, req *http.Request, pa
 		http.NotFound(resp, req)
 	}
 }
+
+// If we've been unable to fetch a fresh OCSP response before expiry of the old
+// one, or, at server start-up, if we're unable to fetch a valid OCSP request at
+// all (either from disk or network), then return false. This signals to the
+// packager that it should not try to package anything; just proxy the content
+// unsigned. (sleevi #8)
+func (this *CertCache) IsHealthy() bool {
+	this.ocspMu.RLock()
+	defer this.ocspMu.RUnlock()
+	return this.isHealthy(this.ocsp)
+}
+
+func (this *CertCache) isHealthy(ocspResp []byte) bool {
+	if ocspResp == nil {
+		log.Println("OCSP response not yet fetched.")
+		return false
+	}
+	issuer := this.findIssuer()
+	if issuer == nil {
+		log.Println("Cannot find issuer certificate in CertFile.")
+		return false
+	}
+	resp, err := ocsp.ParseResponseForCert(ocspResp, this.certs[0], issuer)
+	if err != nil {
+		log.Println("Error parsing OCSP response:", err)
+		return false
+	}
+	if resp.NextUpdate.Before(time.Now()) {
+		log.Println("Cached OCSP is stale, NextUpdate:", resp.NextUpdate)
+		return false
+	}
+	return true
+}
+
+func (this *CertCache) maybeUpdateOCSP() error {
+	this.ocspMu.RLock()
+	var ocspUpdateAfter time.Time
+	ocsp, err := this.ocspFile.Read(context.Background(), this.shouldUpdateOCSP, func(orig []byte) []byte {
+		return this.updateOCSP(orig, &ocspUpdateAfter)
+	})
+	if err != nil {
+		return errors.Wrap(err, "Updating OCSP cache")
+	}
+	if len(ocsp) == 0 {
+		return errors.New("Missing OCSP response.")
+	}
+	if !this.isHealthy(ocsp) {
+		return errors.New("OCSP failed health check.")
+	}
+	// Convert read lock into write lock before updating ocsp and ocspUpdateAfter.
+	this.ocspMuMu.Lock()
+	this.ocspMu.RUnlock()
```

Comment:
In the above failure scenarios, is the read lock unlocked?

Reply:
Good catch. I made this error elsewhere, too (including storage.go). Fixed (as soon as I push).